Retry Philips Hue command when Bridge request failed

domibarton opened the pull request:
As mentioned in #13389, the request to the Hue Bridge can fail,
especially when having a lot of lights and setting a scene for a bunch
of lights.

This is probably an expected behaviour from the Hue Bridge API, as we're
DDoS-ing it with requests. Since these are the limits of the Hue Bridge,
we can workaround that.

This change will introduce retries of the Hue commands. If a command
has failed, the API will no be requested 3 times with a small delay
in between.

## Description:

**Related issue (if applicable):** fixes #13389

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [ ] There is no commented out code in this PR.

domibarton committed:
Retry Philips Hue command when Bridge request failed

As mentioned in #13389, the request to the Hue Bridge can fail,
especially when having a lot of lights and setting a scene for a bunch
of lights.

This is probably an expected behaviour from the Hue Bridge API, as we're
DDoS-ing it with requests. Since these are the limits of the Hue Bridge,
we can workaround that.

This change will introduce retries of the Hue commands. If a command
has failed, the API will no be requested 3 times with a small delay
in between.

homeassistant commented:
Hi @domibarton, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#22645).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @domibarton, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#22645).

Once you do that we will be able to review and accept this pull request.

Thanks!

balloob commented on homeassistant/components/hue/light.py:
```diff
@@ -406,10 +406,27 @@ def device_info(self):
+                    await self.light.set_state(**command)
+                break
+
+            except (RequestError, ResponseError) as ex:
```
When would response error be raised?

domibarton replied:
I had this occuring when I tested it… sometimes the bridge returned an invalid response (text/html) when it was hammered. Didn't saw the content though, as I didn't run HA in debug mode.

Mariusthvdb replied:
If it is of any assistance: received that error yesterday also (text/html) , had not seen it before though.

Error displayed without further info or text for that matter. 

Ha 90.2 here.

domibarton replied:
@balloob I know catching those errors is a bit… generous. But I couldn't find a fast way to differ between an API rate limit error (which results in a `RequestError`) and any other request errors (which also result in a `RequestError`). The same for `ResponseError`. If the error persists over 3 tries, then it will be raised anyway.

What are your opinions here? I just had a deeper look at your  [aiohue](https://github.com/balloob/aiohue) project and here are my thoughts:

It would be much nicer if we identify the specific API rate limit error (probably by the HTTP return code & content) and raise a specific `Exception` in the `aiohue` module. That way, would could catch it in HASS and to exception handling only for this case. 

I also had a look at the exception handling in `aiohue`. It seems that `aiohttp` is raising exceptions with a lot of context. Unfortunately, `aiohue` is wrapping all those exceptions and throwing away all the context (merging all exceptions back into one, removing all informations, not using `raise… from…` syntax properly).

I opened a new [issue in aiohue](https://github.com/balloob/aiohue/issues/17) to address this concern. It will probably help in exactly this case, but if you've a better idea how to handle it, I'm open for it :)

balloob commented on homeassistant/components/hue/light.py:
```diff
@@ -406,10 +406,27 @@ def device_info(self):
+        for i in range(0, 3):
+
+            try:
+                sleep(i * 0.5)
```
Use `asyncio.sleep`

domibarton replied:
thx, replaced it with `asyncio.sleep`

balloob commented on homeassistant/components/hue/light.py:
```diff
@@ -406,10 +406,27 @@ def device_info(self):
+        for i in range(0, 3):
+
+            try:
+                sleep(i * 0.5)
```
Don't sleep if `i == 0`

balloob commented:
I'm a bit hesitant to merge this PR, because it only applies the fix to turn_off, and not turn_on. I wonder how well tested this code is?

balloob commented:
This PR seems to have gone stale. Closing it.

domibarton commented:
@balloob sorry didn't see your last message. I implemented this manually on my HA (RPi) and had to copy-paste the changes to my local editor. You're right, I missed the `async_turn_on` part. I updated it accordingly.

The issue #13389 is still open and I know a bunch of guys which have exactly that problem. I think closing the PR would be wrong, especially if there's no other fix for it.

Never got a response to the aiohue issue, so I didn't had a look here. Is there anyone else who can test the fix?
